### PR TITLE
Core: VehicleJoinEvent crash fix

### PR DESCRIPTION
**Pull request by PolarCookie**
**Changes proposed:**
fix for https://github.com/TrinityCore/TrinityCore/issues/23133

**Issues addressed:** 
#23133

### CODE REVIEW

**Comment by Shauren** [comment]
I'll be blunt here (keep https://github.com/TrinityCore/TrinityCore/issues/23133#issuecomment-476837142 in mind). I hate this fix. I want vehicle auras to be single target again

**Comment by PolarCookie** [comment]
https://github.com/TrinityCore/TrinityCore/pull/22794
Because a lot of vehicle auras are of the type
Effect 1: vehicle control on vehicle
Effect 2: dummy aura on the caster
and this only worked when aura applications were shared, but this caused issues with other multi target spells.

**Comment by PolarCookie** [comment]
And I'll be blunt that it's not a good idea to hack vehicle auras onto a single target system where they don't belong.

**Comment by Shauren** [comment]
They clearly do - there can only be one control vehicle ever

**Comment by PolarCookie** [comment]
They clearly are not flagged as such, and until I see the Blizzard source code I don't assume how they did things.

**Comment by jackpoz** [comment]
any update on this ?

**Comment by PolarCookie** [comment]
As far as I'm concerned it's finished as is; the issue was that `Passenger->SetVehicle(target)` was called while player was already set to a previous vehicle, and that caused issues with the charmer/charmed variables being still set to the previous vehicle later.

File: src/server/game/Entities/Vehicle/Vehicle.cpp [hunk 1]
```diff
@@ -785,6 +785,11 @@ bool VehicleJoinEvent::Execute(uint64, uint32)
         return true;
     }
 
+    //It's possible that multiple vehicle join
+    //events are executed in the same update
```
Comment (illfated) [inline src/server/game/Entities/Vehicle/Vehicle.cpp]: Cosmetic detail, but seeing the length of the code before & after this addition, maybe the comment could be reduced to 1 line?

**Comment by jackpoz** [comment]
wouldn't it be better to do some check when queueing a vehicle join event instead of when executing it ?

**Comment by PolarCookie** [comment]
That's the issue, multiple vehicle join events can get queued up before any are executed, all will gladly pass any test because of this.

**Comment by PolarCookie** [comment]
The test you want is already happening https://github.com/TrinityCore/TrinityCore/blob/f7290509005d3b93bb62149faf3add0bb7f54f51/src/server/game/Entities/Unit/Unit.cpp#L12677-L12686

**Comment by jackpoz** [comment]
Well that doesn't check if there is any join event in the queue. Also is it ok to execute only the first one or should only the last one be executed ?

File: src/server/game/Entities/Vehicle/Vehicle.cpp (hunk 1 above)
Comment (illfated) [reply src/server/game/Entities/Vehicle/Vehicle.cpp]: What I want to suggest here is both to drag back line 2 (789) into line 1 (788) and to add a blank space after the comment backslashes (`//`) like this:
```cpp
    // It's possible that multiple vehicle join events are executed in the same update
```
( I simply thought that this format would follow the code style standard a bit more.)

**Comment by PolarCookie** [comment]
While you're at it, delete `Unit::_EnterVehicle` all together, move all modification done there to  `VehicleJoinEvent::Execute`. Don't think you even need to search the queue that way, would be a significant overhead.

**Comment by jackpoz** [comment]
merging this for now, then we see if there's a better way to handle it

**Comment by jackpoz** [comment]
Thanks for the PR :)

**Pull request merged**